plugins/wasi_nn/neuralspeed.cpp

```diff
@@ -0,0 +1,310 @@
+  if (!Py_IsInitialized()) {
+    spdlog::info(
+        "[WASI-NN][Error] Neural speed backend: Model has been realse, please reload it."sv);
+    return WASINN::ErrNo::RuntimeError;
+  }
+  auto &CxtRef = Env.NNContext[ContextId].get<Context>();
+  auto &GraphRef = Env.NNGraph[CxtRef.GraphId].get<Graph>();
+  if (GraphRef.EnableDebugLog) {
+    spdlog::info("[WASI-NN][Debug] Neural speed backend: compute"sv);
+  }
+  if (CxtRef.Inputs.size() == 0) {
+    spdlog::error("[WASI-NN] Neural speed backend: Llama input is not set!"sv);
+    return ErrNo::InvalidArgument;
+  }
+  CxtRef.Outputs.clear();
+  PyObject *TensorList = PyList_New(0);
+  for (size_t Cnt = 0; Cnt < CxtRef.Inputs.size(); ++Cnt) {
+    PyObject *Num = PyLong_FromLong(CxtRef.Inputs[Cnt]);
+    PyList_Append(TensorList, Num);
+    Py_DECREF(Num);
+  }
+  PyObject *TmpArg = PyList_New(0);
+  PyList_Append(TmpArg, TensorList);
+  PyObject *TorchModule = PyImport_ImportModule("torch");
+  if (TorchModule == nullptr) {
+    spdlog::error(
+        "[WASI-NN] neural speed backend: Can not find torch library."sv);
+    return WASINN::ErrNo::RuntimeError;
+  }
+  PyObject *LongTensorFunc = PyObject_GetAttrString(TorchModule, "LongTensor");
+  PyObject *LongTensorArgs = PyTuple_Pack(1, TmpArg);
+  PyObject *LongTensor = PyObject_CallObject(LongTensorFunc, LongTensorArgs);
+  Py_DECREF(TensorList);
+  Py_DECREF(TorchModule);
+  Py_DECREF(LongTensorFunc);
+  Py_DECREF(TmpArg);
+  Py_DECREF(LongTensorArgs);
+  if (LongTensor == nullptr) {
+    spdlog::error(
+        "[WASI-NN] neural speed backend: Input transfer tensor failed."sv);
+    return WASINN::ErrNo::InvalidArgument;
+  }
+  // PyObject *GenerateArgs = PyTuple_Pack(1, LongTensor);
+  PyObject *Result = PyObject_CallMethodObjArgs(
+      GraphRef.Model, PyUnicode_FromString("generate"), LongTensor, NULL);
+  if (Result == nullptr) {
+    PyErr_Print();
+    spdlog::error(
+        "[WASI-NN] neural speed backend: Neural Speed runtime error."sv);
+    return WASINN::ErrNo::RuntimeError;
+  }
+  if (PyList_Check(Result)) {
+    const Py_ssize_t OuterSize = PyList_Size(Result);
+    for (Py_ssize_t OutterCnt = 0; OutterCnt < OuterSize; ++OutterCnt) {
+      PyObject *InnerList = PyList_GetItem(Result, OutterCnt);
+      if (PyList_Check(InnerList)) {
+        std::vector<long long int> InnerVec;
+        const Py_ssize_t InnerSize = PyList_Size(InnerList);
+        for (Py_ssize_t InnerCnt = 0; InnerCnt < InnerSize; ++InnerCnt) {
+          PyObject *Num = PyList_GetItem(InnerList, InnerCnt);
+          if (PyLong_Check(Num)) {
+            InnerVec.push_back(PyLong_AsLong(Num));
+          }
+          Py_DECREF(Num);
+        }
+        CxtRef.Outputs = InnerVec;
+      }
+      Py_DECREF(InnerList);
+    }
+  }
+  Py_DECREF(Result);
+  // Py_DECREF(GenerateArgs);
+  Py_DECREF(LongTensor);
+  return WASINN::ErrNo::Success;
+}
+
+Expect<WASINN::ErrNo> finiSingle(WASINN::WasiNNEnvironment &Env,
```

Comment:
`finiSingle` is used to clear the context after `computeSingle` has been executed. If `computeSingle` is not used, implementing `finiSingle` is not necessary; implementing `unload` should be sufficient.

Reply:
Since I am not sure how to call unload from `wasmedge-wasi-nn`, I used `finiSingle` to replace it.
Can you tell me how to use `wasmedge-wasi-nn` call the unload function? Thank you.

Reply:
The `unload()` function was recently added to `wasmedge-wasi-nn` in version `0.7.1`. You can check the example [here](https://github.com/second-state/WasmEdge-WASINN-examples/blob/master/wasmedge-ggml/test/unload/src/main.rs#L100).

Reply:
I have fixed it. Please check thx.